### PR TITLE
[2.7] Add deferrable config to the Deferrable operators guide

### PR DESCRIPTION
**Pull request by TJaniF**
Ready to be merged :) 

-----------

Resolves: #2671

### CODE REVIEW

File: learn/deferrable-operators.md [hunk 1]
```diff
@@ -94,14 +118,14 @@ There are numerous benefits to using deferrable operators including:
...
-- There are some use cases where it can be more appropriate to use a traditional sensor instead of a deferrable operator. For example, if your task needs to wait only a few seconds for a condition to be met, Astronomer recommends using a Sensor in [`reschedule` mode](https://github.com/apache/airflow/blob/1.10.2/airflow/sensors/base_sensor_operator.py#L46-L56) to avoid unnecessary resource overhead.
+- In situations where you cannot use a deferrable operator for a longer running sensor task, for example because your infrastructure does not support a triggerer, Astronomer recommends using a sensor in [`reschedule` mode](https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/sensors.html) to reduce unnecessary resource overhead. See the [Airflow documentation](https://airflow.apache.org/docs/apache-airflow/stable/authoring-and-scheduling/deferring.html#difference-between-mode-reschedule-and-deferrable-true-in-sensors) for details about the differences between `reschedule` mode and deferrable operators.
```
Comment (TJaniF) [inline learn/deferrable-operators.md:121]: I am on the fence here between linking to the OSS docs and repeating what they say in a short section below. Something like 

### Reschedule vs deferrable

🤔

File: learn/deferrable-operators.md [hunk 2]
```diff
@@ -52,11 +52,35 @@ If you are not using Astro, run `airflow triggerer` to start a triggerer process
...
 
-### Update your DAGs
+### Enable deferrable mode
 
-To use a deferrable version of a core Airflow operator in your DAG, you only need to replace the import statement for the existing operator. For example, Airflow's `TimeSensorAsync` is a replacement of the non-deferrable `TimeSensor` ([source](https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/sensors/time_sensor/index.html?highlight=timesensor#module-contents)). To use `TimeSensorAsync`, remove your existing `import` and replace it with the following:
+Many operators Airflow operators (e.g. the [TriggerDagRunOperator](https://registry.astronomer.io/providers/apache-airflow/versions/latest/modules/TriggerDagRunOperator) or the [WasbBlobSensor](https://registry.astronomer.io/providers/apache-airflow-providers-microsoft-azure/versions/latest/modules/WasbBlobSensor)) can be set to run in deferrable mode using the `deferrable` parameter. You can learn if the operator you want to use has a `deferrable` parameter by looking at the list of operator parameters in the [Astronomer Registry](https://registry.astronomer.io/).
```
Comment (kentdanas) [inline learn/deferrable-operators.md:59]:
```suggestion
Many Airflow operators (e.g. the [TriggerDagRunOperator](https://registry.astronomer.io/providers/apache-airflow/versions/latest/modules/TriggerDagRunOperator) or the [WasbBlobSensor](https://registry.astronomer.io/providers/apache-airflow-providers-microsoft-azure/versions/latest/modules/WasbBlobSensor)) can be set to run in deferrable mode using the `deferrable` parameter. You can learn if the operator you want to use has a `deferrable` parameter by looking at the list of operator parameters in the [Astronomer Registry](https://registry.astronomer.io/).
```

File: learn/deferrable-operators.md [hunk 3]
```diff
@@ -52,11 +52,35 @@ If you are not using Astro, run `airflow triggerer` to start a triggerer process
...
+
+To set all tasks defined with an operator that has a `deferrable` parameter in your whole Airflow instance to use deferrable mode, set the Airflow config `operators.default_deferrable` to `True`. You can do so by defining the following environment variable in your Airflow environment:
```
Comment (kentdanas) [inline learn/deferrable-operators.md:61]:
```suggestion
To set all tasks defined with an operator that has a `deferrable` parameter in your Airflow instance to use deferrable mode, set the Airflow config `operators.default_deferrable` to `True`. You can do so by defining the following environment variable in your Airflow environment:
```

File: learn/deferrable-operators.md [hunk 4]
```diff
@@ -52,11 +52,35 @@ If you are not using Astro, run `airflow triggerer` to start a triggerer process
...
+   trigger_dag_id="helper_dag_wait_30_seconds",
+   wait_for_completion=True,
+   poke_interval=20,
+   deferrable=True,  # turns on deferrable mode just for this operator instance
+)
+```
+
+### Replace existing operators with deferrable versions
+
+If the operator you want to use does not have a `deferrable` parameter you can look for a deferrable version of the operator in the [Astronomer Registry](https://registry.astronomer.io/). These operators usually have the same name as the original operator, but with the word `Async` appended to the end. For example, the [TimeSensorAsync](https://registry.astronomer.io/providers/apache-airflow/versions/latest/modules/TimeSensorAsync) is the deferrable version of the [TimeSensor](https://registry.astronomer.io/providers/apache-airflow/versions/latest/modules/TimeSensor).
```
Comment (kentdanas) [inline learn/deferrable-operators.md:81]:
```suggestion
If the operator you want to use does not have a `deferrable` parameter, you can look for a separate deferrable version of the operator in the [Astronomer Registry](https://registry.astronomer.io/). These operators usually have the same name as the original operator, but with the word `Async` appended to the end. For example, the [TimeSensorAsync](https://registry.astronomer.io/providers/apache-airflow/versions/latest/modules/TimeSensorAsync) is the deferrable version of the [TimeSensor](https://registry.astronomer.io/providers/apache-airflow/versions/latest/modules/TimeSensor).
```

File: learn/deferrable-operators.md [hunk 5]
```diff
@@ -94,14 +118,14 @@ There are numerous benefits to using deferrable operators including:
 Some additional notes about using deferrable operators:
```
Comment (kentdanas) [inline learn/deferrable-operators.md:118]: With the new sections, it seems weird to me to have these bulleted lists in the "replace existing operators section". Maybe the benefits bullets can be moved under Terms and Concepts? And "additional notes" another subsection under "use deferrable operators"?

File: learn/deferrable-operators.md [hunk 6]
```diff
@@ -94,14 +118,14 @@ There are numerous benefits to using deferrable operators including:
...
 
 - If you want to replace non-deferrable operators in an existing project with deferrable operators, Astronomer recommends importing the deferrable operator class as its non-deferrable class name. If you don't include this part of the import statement, you need to replace all instances of non-deferrable operators in your DAGs. In the above example, that would require replacing all instances of `TimeSensor` with `TimeSensorAsync`.
```
Comment (kentdanas) [inline learn/deferrable-operators.md:120]:
```suggestion
```
Actually, related to my comment below, let's delete this bullet (less and less relevant with the param, and people can figure this out themselves). And then find a place for your bullet below.

File: learn/deferrable-operators.md (hunk 1 above)
Comment (kentdanas) [reply learn/deferrable-operators.md:121]: I think linking to Airflow docs here is fine. Given I deleted the previous bullet and this is a short note, maybe we pull it out into a highlighted note?

**Review by kentdanas — APPROVED** [review]
Looks great. Multiple (maybe confusing, sorry) comments on what we should do with the bulleted lists, but otherwise I think we're good here. Let me know if you want to chat about any of that.

File: learn/deferrable-operators.md (hunk 2 above)
Comment (TJaniF) [reply learn/deferrable-operators.md]: 🤦

**Review by TJaniF — COMMENTED** [review]
@kentdanas thank you! :) 

@jwitz and another small one :)

File: learn/what-is-a-sensor.md [hunk 7]
```diff
@@ -133,6 +133,12 @@ When using sensors, keep the following in mind to avoid potential performance is
...
 
-[Deferrable operators](deferrable-operators.md) (sometimes referred to as asynchronous operators) were released with Airflow 2.2 and are designed to eliminate the problem of any operator or sensor taking up a full worker slot for the entire time they are running. Deferrable versions of many sensors exist in open source Airflow and in the [Astronomer Providers package](https://github.com/astronomer/astronomer-providers). Astronomer recommends using these in most cases to reduce resource costs.
+[Deferrable operators](deferrable-operators.md) (sometimes referred to as asynchronous operators) were released with Airflow 2.2 and are designed to eliminate the problem of any operator or sensor taking up a full worker slot for the entire time they are running. Many operators have a `deferrable` parameter which can be set to `True` to make the operator deferrable. For many sensors where this parameter is not available, deferrable versions exist in open source Airflow and in the [Astronomer Providers package](https://github.com/astronomer/astronomer-providers) (often having `Async` appended to the operator name). Astronomer recommends using these in most cases to reduce resource costs.
```
Comment (jwitz) [inline learn/what-is-a-sensor.md:136]:
```suggestion
[Deferrable operators](deferrable-operators.md) (sometimes referred to as asynchronous operators) were released with Airflow 2.2 and are designed to eliminate the problem of any operator or sensor taking up a full worker slot for the entire time they are running. Many operators have a `deferrable` parameter which can be set to `True` to make the operator deferrable. For the sensors where this parameter is not available, deferrable versions exist in open source Airflow and in the [Astronomer Providers package](https://github.com/astronomer/astronomer-providers). Astronomer recommends using these in most cases to reduce resource costs.
```

File: learn/what-is-a-sensor.md [hunk 8]
```diff
@@ -133,6 +133,12 @@ When using sensors, keep the following in mind to avoid potential performance is
...
 
-For DAG authors, using deferrable sensors is no different from using regular sensors. All you need is to do is run a `triggerer` process in Airflow and replace the names of all sensors in your DAG code with their deferrable counterparts.  For more details, see [Deferrable operators](deferrable-operators.md). 
+For DAG authors, using deferrable sensors is no different from using regular sensors. All you need is to do is run a `triggerer` process in Airflow and either:
+
+- Set the Airflow config `operators.default_deferrable` to `True` to set all sensors with a `deferrable` parameter to use deferrable mode by default.
```
Comment (jwitz) [inline learn/what-is-a-sensor.md:140]:
```suggestion
- Set the Airflow config `operators.default_deferrable` to `True` to set all sensors with a `deferrable` parameter to be deferrable by default.
```

**Review by jwitz — APPROVED** [review]

File: learn/deferrable-operators.md [hunk 9]
```diff
@@ -84,24 +112,19 @@ If you are using a deferrable operator that is part of a provider package, you w
    )
    ```
 
-Note that importing the asynchronous operator using the alias of the analogous traditional operator (e.g. `import SnowflakeOperatorAsync as SnowflakeOperator`) is simply to make updating existing DAGs easier. This is not required, and may not be preferrable when authoring a new DAG.
+Note that importing the asynchronous operator using the alias of the analogous traditional operator (e.g. `import SnowflakeOperatorAsync as SnowflakeOperator`) is simply to make updating existing DAGs easier. This is not required, and may not be preferable when authoring a new DAG.
 
 There are numerous benefits to using deferrable operators including:
```
Comment (kentdanas) [inline learn/deferrable-operators.md:117]: @TJaniF let's move these bullets up to the Terms and Concepts section. I find them weird here in a section on implementation.

**Review by TJaniF — COMMENTED** [review]
Thank you @jwitz :)

File: learn/deferrable-operators.md [hunk 10]
```diff
@@ -52,11 +58,35 @@ If you are not using Astro, run `airflow triggerer` to start a triggerer process
...
+trigger_dag_run = TriggerDagRunOperator(
+   task_id="trigger_dag_run",
+   trigger_dag_id="helper_dag_wait_30_seconds",
+   wait_for_completion=True,
+   poke_interval=20,
+   deferrable=True,  # turns on deferrable mode just for this operator instance
+)
+```
+
+### Replace existing operators with deferrable versions
```
[the remark this answers is not in the document]
Comment (kentdanas) [reply learn/deferrable-operators.md]: Yep, makes sense to me!

**Comment by TJaniF** [comment]
And this one is ready as well :)

**Pull request merged**